Make all security annotations inherited; correct how Arc processes non-inherited interceptor bindings

manovotn opened the pull request:
Fixes #33666 

ATM Arc assumes that all class level bindings are inherited, which is also recommended by CDI.
However, once you start using `strict mode` switch, we properly check bindings and only inherited those with respective annotation. In such a case, Quarkus security stops working as intended.

Feel free to browse #33666 for more details on this.

This PR:
1. Changes security annotations, which Quarkus marks as interceptor bindings, to always appear as `@Inherited`.
2. Changes Arc to now process non-inherited interceptor bindings properly even outside of strict compatibility mode

Note that the latter change (correction to our behavior, really) might break some existing apps who had hierarchical intercepted structures and relied on inheritance without having their bindings `@Inherited`.

I also tried to add a simple test showing/testing this for all security annotations except `@PermitAll` where I couldn't figure out a case where it would matter (feel free to enlighten me, I am a security noob :)).

manovotn committed:
Transform security annotations to be inherited; add tests

manovotn committed:
Arc - treat non-inherited interceptor bindings properly even outside strict mode

Ladicek commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:
```diff
@@ -932,7 +932,7 @@ private void doAddClassLevelBindings(ClassInfo classInfo, Collection<AnnotationI
             if (superClass != null) {
                 // proper interceptor binding inheritance only in strict mode, due to Quarkus expecting security
                 // annotations (such as `@RolesAllowed`) to be inherited, even though they are not `@Inherited`
```
The 2 lines above need to be deleted now.

manovotn replied:
Good point, done.

Ladicek reviewed (approved):
LGTM otherwise. Thanks!

michalvavrik commented on extensions/security/deployment/src/test/java/io/quarkus/security/test/cdi/inheritance/SubclassDenyAllBean.java:
```diff
@@ -0,0 +1,7 @@
+
+@ApplicationScoped
+public class SubclassDenyAllBean extends DenyAllBean {
+}
```
@manovotn I'd like to check I get it correctly. Here, we need `@Inherited` to have secured `ping` method declared on `DenyAllBean` that has a class-level `DenyAll` annotation? And if you added `pong` to `SubclassDenyAllBean` this still won't be secured based on `DenyAll` placed on the parent, correct?

that's what I expect, I'd just like to understand this: without `@Inherited` the `ping` method would not be secured, do I get it right?

Ladicek replied:
That's correct, yes.

michalvavrik replied:
Cool, in that case changes indeed keep current behavior. Thanks.

sberyozkin replied:
@michalvavrik thanks, let me check

manovotn replied:
Ladislav beat me to it but yes, that's exactly the case.
I think I saw some other tests that verified this very behavior elsewhere so my tests focused just on the inheritance.

manovotn commented:
> > @permitAll where I couldn't figure out a case where it would matter (feel free to enlighten me, I am a security noob :)).
> 
> As far as class-level `@PermitAll` goes, we have some flags that deny unsecured endpoints and `@PermitAll` is considered a security annotation so it allows you to avoid `deny-all` or `roles-allowed` default JAX-RS security. We have also one property for unannotated beans but I never touched.

Hm, OK. Not sure it's worth adding JAX-RS test here; basically having the rest of annotations tested means this one will be inherited as well :shrug:

michalvavrik commented:
> Hm, OK. Not sure it's worth adding JAX-RS test here; basically having the rest of annotations tested means this one will be inherited as well 🤷

it isn't mainly because JAX-RS skips CDI interceptors :-D I was just happy I new the answer. Your tests are ok, we have extensive test coverage, higher hundreds, I am not worried if CI is green.

sberyozkin commented:
Thanks, making security annotations inheritable should only improve things and also act as a hardening fix where users may think putting a non-inheritable annotation on a super abstract class is enough to cover the whole hierarchy

Ladicek commented:
I'll look into the MP FT TCK failures, those are strange.

michalvavrik commented:
> Thanks, making security annotations inheritable should only improve things and also act as a hardening fix where users may think putting a non-inheritable annotation on a super abstract class is enough to cover the whole hierarchy

I still believe it is not hardening as this PR should left security behavior unchanged. @manovotn @Ladicek ?

Ladicek commented:
Yeah, exactly. We're preserving the current behavior here. So far, the annotations have been treated as inherited even though they are not `@Inherited`. We're just making them `@Inherited` (for the purpose of CDI).

manovotn commented:
> I'll look into the MP FT TCK failures, those are strange.

These are related and the culprit is how SR FT extension applies [`FaultToleranceBinding`](https://github.com/smallrye/smallrye-fault-tolerance/blob/main/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/FaultToleranceBinding.java) - a binding that's *not* marked as `@Inherited` which causes the failures.
Should be fixable on Quarkus or SR FT side, looking into what's the best option now.

Ladicek committed:
ArC: fix non-@Inherited transitive bindings on @Inherited bindings

If a non-`@Inherited` binding A is present on an `@Inherited` binding B,
A would previously be ignored when B is processed in an `onlyInherited`
context. This is wrong: transitive bindings don't have to be `@Inherited`
to apply on a class that inherits a "direct" binding.

This commit fixes that by moving the "ignore non-@Inherited bindings" logic
to the `BeanDeployment.extractInterceptorBindings()` method, which can
distinguish between "direct" bindings (`@Inherited` may apply) and
transitive bindings (`@Inherited` never applies).

Ladicek commented:
Pushed a commit that should fix the MP FT TCK.

michalvavrik reviewed (approved):
I played with it little bit. LGTM.
